### ISSUE TITLE
[!!!][TASK] Remove FlexFormTools traverser

### ISSUE THREAD

**Issue by TYPO3IncTeam** [TYPO3-Documentation/Changelog-To-Doc#706]
:information_source: View this commit [on Github](https://github.com/TYPO3/typo3/commit/abc7662b09831e5e61dcfe3a01aee2db6df02a07)
:busts_in_silhouette: Authored by Christian Kuhn lolli@schwarzbu.ch
:heavy_check_mark: Merged by Benni Mack benni@typo3.org

## Commit message

[!!!][TASK] Remove FlexFormTools traverser

With various pre-patches being done, the ugly traverser
method FlexFormTools->traverseFlexFormXMLData() is
obsolete.

This final patch to FlexFormTools removes that method
and its helper from the codebase. We're actively
removing the method and skip a deprecation phase in
this case: First, the method is low level and usually
not used by extension developers. Extensions like
the TemplaVoila forks have their own implementations
of FlexFormTools already, their migration is straight
forward. Secondly, we probably want to further simplify
FlexForm related details, maintaining the old traverser
would be a hindrance since it's so hard to follow.

FlexFormTools is now a stateless service. The patch
comes with a .rst file mentioning the changes, and
documents some details from previous patches.

Resolves: #102229
Related: #102218
Related: #102212
Releases: main
Change-Id: I866e8c92c07753b8c53ae3d8233c55d35614cb85
Reviewed-on: https://review.typo3.org/c/Packages/TYPO3.CMS/+/81504
Tested-by: Stefan Bürk <stefan@buerk.tech>
Reviewed-by: Benni Mack <benni@typo3.org>
Tested-by: core-ci <typo3@b13.com>
Reviewed-by: Stefan Bürk <stefan@buerk.tech>
Tested-by: Benni Mack <benni@typo3.org>

## :heavy_plus_sign: Added files

<details>
<summary>13.0/Breaking-102229-RemovedFlexFormTools-traverseFlexFormXMLData.rst</summary>

```rst
.. include:: /Includes.rst.txt

.. _breaking-102229-1698053674:

====================================================================
Breaking: #102229 - Removed FlexFormTools->traverseFlexFormXMLData()
====================================================================

See :issue:`102229`

Description
===========

Class :php:`TYPO3\CMS\Core\Configuration\FlexForm\FlexFormTools` got a series
of cleanup and removal patches.

The following public class properties have been removed:

* :php:`$reNumberIndexesOfSectionData`
* :php:`$flexArray2Xml_options`
* :php:`$callBackObj`
* :php:`$cleanFlexFormXML`

The following public methods have been removed:

* :php:`traverseFlexFormXMLData()`
* :php:`traverseFlexFormXMLData_recurse()`
* :php:`cleanFlexFormXML_callBackFunction()`

The following public methods have been marked `@internal`:

* :php:`cleanFlexFormXML()`
* :php:`flexArray2Xml()`
* :php:`migrateFlexFormTcaRecursive()`

The class is now a stateless service and can be injected as shared service
without any risk of triggering side effects.


Impact
======

In general, these changes should have relatively low impact on extensions, if they
don't built additional low level functionality on top of the general TYPO3 core
FlexForm related features. Extensions like the TemplaVoila forks may need to have
a look for required adaptions, though.

Using the removed methods or properties in TYPO3 v13 will of course trigger PHP
fatal errors.


Affected installations
======================

Instances that extend functionality of FlexForm handling may be affected if they
use methods of class :php:`FlexFormTools`. This is a relatively rare case, most
instances will not be affected when they just provide and use casual FlexForm
definitions in extensions.

The extension scanner will find possible extensions that consume the methods or
properties as a weak match.


Migration
=========

If at all, method :php:`traverseFlexFormXMLData()` is probably the one used in
extensions. The easiest way is to copy the method and it's recursive worker method
to an own class.

Extension developers are however encouraged to refactor their code since
:php:`traverseFlexFormXMLData()` with its callback logic was ugly, hard to follow
and maintain. The core switched away from the method by implementing own traversers
that match the specific use cases. Method :php:`cleanFlexFormXML()` is an
example of such an implementation. Note FlexForms are *not* recursive since
since section containers can not be nested since TYPO3 v8 anymore. The core thus
uses some nested foreach loops instead of a recursive approach.


.. index:: FlexForm, PHP-API, FullyScanned, ext:core

```

</details>

**Comment by brotkrueml** [comment]
See: https://docs.typo3.org/m/typo3/reference-coreapi/12.4/en-us/ApiOverview/FlexForms/Index.html#how-to-modify-flexforms-from-php
